bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,236 @@
+    /// This update will only contain [`Txid`]s in SparseChain, and no actual transaction data.
+    ///
+    /// User needs to fetch the required transaction data and update the [`ChainGraph`] before applying it.
+    pub fn spk_txid_scan(
+        &self,
+        spks: impl Iterator<Item = Script>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+    ) -> Result<SparseChain, ElectrumError> {
+        let mut dummy_keychains = BTreeMap::new();
+        dummy_keychains.insert((), spks.enumerate().map(|(i, spk)| (i as u32, spk)));
+
+        Ok(self.wallet_txid_scan(dummy_keychains, None, local_chain)?.0)
+    }
+
+    /// Scan for a keychain tracker, and create an initial [`KeychainScan`] candidate update.
+    /// This update will only contain [`Txid`]s in SparseChain, and no actual transaction data.
+    ///
+    /// User needs to fetch the required transaction data and update the [`KeychainScan`] before applying it.
+    pub fn wallet_txid_scan<K: Ord + Clone>(
+        &self,
+        scripts: BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        stop_gap: Option<usize>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+    ) -> Result<(SparseChain, BTreeMap<K, u32>), ElectrumError> {
+        let mut sparse_chain = SparseChain::default();
+
+        // Check for reorgs.
+        // In case of reorg, new checkpoints until the last common checkpoint is added to the structure
+        for (&existing_height, &existing_hash) in local_chain.iter().rev() {
+            let current_hash = self
+                .client
+                .block_header(existing_height as usize)?
+                .block_hash();
+            sparse_chain
+                .insert_checkpoint(BlockId {
+                    height: existing_height,
+                    hash: current_hash,
+                })
+                .expect("This never errors because we are working with a fresh chain");
+
+            if current_hash == existing_hash {
+                break;
+            }
+        }
+
+        // Insert the new tip
+        let (tip_height, tip_hash) = self.get_tip()?;
+        sparse_chain
+            .insert_checkpoint(BlockId {
+                height: tip_height,
+                hash: tip_hash,
+            })
+            .expect("This never errors because we are working with a fresh chain");
+
+        let mut keychain_index_update = BTreeMap::new();
+
+        let mut reorgred_tx = Vec::new();
+
+        // Fetch Keychain's last_active_index and all related txids.
+        // Add them into the KeyChainScan
+        for (keychain, mut scripts) in scripts.into_iter() {
+            let mut last_active_index = 0;
+            let mut unused_script_count = 0usize;
+            let mut script_history_txid = Vec::<(Txid, TxHeight)>::new();
+
+            loop {
+                if let Some((index, script)) = scripts.next() {
+                    let history = self
+                        .script_get_history(&script)?
```

Comment:
shouldn't we be using `batch_script_get_history` for efficiency here?

Reply:
We can, but currently we are getting all the scripts using `iter_all_script_pubkeys_by_keychain()`, which calls  `descriptor_into_script_iter()` which is an iterator over all the scripts that can be derived from a descriptor. So if we call `batch_script_history`, it will work for the `Sync` call, but for `Scan` it will try to fetch for `2^32` values for wild card descriptors which can be a pretty massive call..

Reply:
Sorry I missed that comment. Look at what we do in electrum impl in bdk already. You just take a bunch of scripts up to `batch_size` which the caller can pass in and batch it in that way.